### ISSUE TITLE
Custom variant of YoloV8n-seg

### ISSUE THREAD

**Issue by imvamsidhark** [ultralytics/ultralytics#19263]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

I’m working on a variant of yolov8n-seg model where it gives 10 tensors instead of two tensors (output0, output1), it gives the following tensors with the shapes
 
bbox_0 shape: (1, 4, 80, 80) 
bbox_1 shape: (1, 4, 40, 40) 
bbox_2 shape: (1, 4, 20, 20)
class_prob_0 shape: (1, 10, 80, 80) 
class_prob_1 shape: (1, 10, 40, 40)
class_prob_2 shape: (1, 10, 20, 20)
mask_coeff_0 shape: (1, 32, 80, 80)
mask_coeff_1 shape: (1, 32, 40, 40)
mask_coeff_2 shape: (1, 32, 20, 20)
mask shape: (1, 32, 160, 160) which is mask data

So how to post process this information, generally what’s the approach, moreover I’m working in cpp where I have the tensor file which’s dumped into a binary file (possibly) and I want to extract boxes and segmentation mask out of that binary file.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @imvamsidhark, thank you for your interest in Ultralytics 🚀! We recommend exploring the [Docs](https://docs.ultralytics.com/) for a comprehensive overview of our models and functionality. If you're new to YOLOv8, you can find useful resources for [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage that may help with your work.

For your query involving a custom model variant and post-processing output tensors, we encourage you to provide additional details or share a 🐛 [minimum reproducible example (MRE)](https://docs.ultralytics.com/help/minimum-reproducible-example/) if applicable. This would greatly assist us in replicating your scenario and offering more precise guidance.

If your question involves C++ operations and binary file dumping, please share the following:
- A brief sample of your tensor file or its structure.
- Steps you've taken so far to extract bounding boxes and masks.
- Any errors or challenges you've encountered.

You can also ensure you’re using the latest features by upgrading your package:
```bash
pip install -U ultralytics
```

### Community and Support
While awaiting assistance from our engineers, you can ask further questions or seek advice in our supportive community:
- 💬 Chat: [Discord](https://discord.com/invite/ultralytics)
- 🗣️ Forum: [Discourse](https://community.ultralytics.com/)
- 📚 Subreddit: [r/Ultralytics](https://reddit.com/r/Ultralytics)

For running YOLOv8 in common environments, check the free GPU-enabled setups below:
- **Notebooks**: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Docker**: <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

### Status
<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a> If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are passing, ensuring smooth performance for all supported modes and tasks.

This is an automated response 🛠️, but an Ultralytics engineer will review your issue and provide guidance soon. Thanks for being part of the community and for exploring YOLOv8's possibilities! 🚀

**Comment by Y-T-G** [comment]
You concatenate the bbox, class probabilities and coefficients and flatten the last two dims, which turns them into

bbox: `(1, 4, 8400)`
probs: `(1, 10, 8400)`
coeffs: `(1, 32, 8400)`.

Then you reshape them to `(1, 8400, N)`, get max probs for each of the 8400 output, remove all the outputs less than threshold and then apply NMS to the remaining.

The coefficient is matrix multiplied with mask to get a mask of 160x160 for each object.

**Comment by imvamsidhark** [comment]
```python
import numpy as np
import torch
import torchvision
```

### Loading Binary

```python
def load_preds_from_bin(bin_path):
    """
    Loads predictions from a binary file into PyTorch tensors.
    
    Expected tensor order and shapes:
        bbox_0:        (1, 4, 80, 80)
        bbox_1:        (1, 4, 40, 40)
        bbox_2:        (1, 4, 20, 20)
        class_prob_0:  (1, 10, 80, 80)
        class_prob_1:  (1, 10, 40, 40)
        class_prob_2:  (1, 10, 20, 20)
        mask_coeff_0:  (1, 32, 80, 80)
        mask_coeff_1:  (1, 32, 40, 40)
        mask_coeff_2:  (1, 32, 20, 20)
        mask_data:     (1, 32, 160, 160)
    
    Total elements: 1,205,600 (each 4 bytes in FP32)
    """
    # Load all data from the binary file as a NumPy array.
    data = np.fromfile(bin_path, dtype=np.float32)
    expected_total = 1205600
    assert data.size == expected_total, f"Expected {expected_total} elements, got {data.size}"
    
    offset = 0
    def get_tensor(shape):
        nonlocal offset
        size = np.prod(shape)
        # Convert the appropriate slice of the data into a PyTorch tensor.
        tensor = torch.from_numpy(data[offset:offset + size].reshape(shape))
        offset += size
        return tensor
    
    # Create PyTorch tensors for each prediction.
    bbox_0       = get_tensor((1, 4, 80, 80))
    bbox_1       = get_tensor((1, 4, 40, 40))
    bbox_2       = get_tensor((1, 4, 20, 20))
    class_prob_0 = get_tensor((1, 10, 80, 80))
    class_prob_1 = get_tensor((1, 10, 40, 40))
    class_prob_2 = get_tensor((1, 10, 20, 20))
    mask_coeff_0 = get_tensor((1, 32, 80, 80))
    mask_coeff_1 = get_tensor((1, 32, 40, 40))
    mask_coeff_2 = get_tensor((1, 32, 20, 20))
    mask_data    = get_tensor((1, 32, 160, 160))
    
    return (bbox_0, bbox_1, bbox_2,
            class_prob_0, class_prob_1, class_prob_2,
            mask_coeff_0, mask_coeff_1, mask_coeff_2,
            mask_data)

# Usage example:
bbox_0, bbox_1, bbox_2, class_prob_0, class_prob_1, class_prob_2, mask_coeff_0, mask_coeff_1, mask_coeff_2, mask_data = load_preds_from_bin("10_tensors_data")
```

```python
# Flatten the spatial dimensions for each bbox tensor
bbox_0_flat = bbox_0.view(1, 4, -1)  # Shape: (1, 4, 6400)
bbox_1_flat = bbox_1.view(1, 4, -1)  # Shape: (1, 4, 1600)
bbox_2_flat = bbox_2.view(1, 4, -1)  # Shape: (1, 4, 400)

# Concatenate along the flattened dimension
bbox = torch.cat([bbox_0_flat, bbox_1_flat, bbox_2_flat], dim=2)  # Shape: (1, 4, 8400)

# Similarly, flatten the class probability
class_prob_0_flat = class_prob_0.view(1, 10, -1)  # Shape: (1, 10, 6400)
class_prob_1_flat = class_prob_1.view(1, 10, -1)  # Shape: (1, 10, 1600)
class_prob_2_flat = class_prob_2.view(1, 10, -1)  # Shape: (1, 10, 400)

# Concatenate along the flattened dimension
class_prob = torch.cat([class_prob_0_flat, class_prob_1_flat, class_prob_2_flat], dim=2)  # Shape: (1, 10, 8400)

# Similarly, flatten the mask coefficients
mask_coeff_0_flat = mask_coeff_0.view(1, 32, -1)  # Shape: (1, 32, 6400)
mask_coeff_1_flat = mask_coeff_1.view(1, 32, -1)  # Shape: (1, 32, 1600)
mask_coeff_2_flat = mask_coeff_2.view(1, 32, -1)  # Shape: (1, 32, 400)

# Concatenate along the flattened dimension
mask_coeff = torch.cat([mask_coeff_0_flat, mask_coeff_1_flat, mask_coeff_2_flat], dim=2)  # Shape: (1, 32, 8400)

# The mask_data tensor is already in the correct shape
mask_data = mask_data  # Shape: (1, 32, 160, 160)
```

```python
# Reshape them to (1, 8400, N)
bbox = bbox.permute(0, 2, 1)  # Shape: (1, 8400, 4)
class_prob = class_prob.permute(0, 2, 1)  # Shape: (1, 8400, 10)
mask_coeff = mask_coeff.permute(0, 2, 1)  # Shape: (1, 8400, 32)
```

```python
# Get max probs and class ids for each of the 8400 outputs
max_probs, class_ids = torch.max(class_prob[:, :, 1:], dim=2)  # Shape: (1, 8400)

# Remove predictions with confidence below a threshold
conf_threshold = 0.4
mask = max_probs > conf_threshold
bbox_filtered = bbox[mask]
scores_filtered = max_probs[mask]
class_ids_filtered = class_ids[mask]
mask_coeff_filtered = mask_coeff[mask]

# Apply NMS to the filtered predictions
iou_threshold = 0.45
indices = torchvision.ops.nms(bbox_filtered, scores_filtered, iou_threshold)
```

```python
# Final filtered predictions
bboxes = bbox_filtered[indices] # Shape: (N, 4)
scores = scores_filtered[indices] # Shape: (N,)
class_ids = class_ids_filtered[indices] # Shape: (N,)
mask_coeffs = mask_coeff_filtered[indices] # Shape: (N, 32)
```

```python
# Process the mask coefficients to generate instance masks
mask_proto = mask_data[0]  # Shape: (32, 160, 160)

# Flatten the dimensions of mask_proto
mask_proto_flat = mask_proto.view(32, -1)  # Shape: (32, 25600)

# Final masks
masks = torch.matmul(mask_coeffs, mask_proto_flat)  # Shape: (N, 25600)

# Reshape the masks to (N, 160, 160)
masks = masks.view(-1, 160, 160)

# Apply Sigmoid activation to get the final masks
masks = torch.sigmoid(masks)
```

```python
# Convert torch tensors to NumPy arrays if needed.
if isinstance(bboxes, torch.Tensor):
    bboxes = bboxes.cpu().numpy()
if isinstance(scores, torch.Tensor):
    scores = scores.cpu().numpy()
if isinstance(class_ids, torch.Tensor):
    class_ids = class_ids.cpu().numpy()
if isinstance(masks, torch.Tensor):
    masks = masks.cpu().numpy()

# --- Load the source image ---
image = cv2.imread("source.png")
if image is None:
    raise FileNotFoundError("Could not load source.png")
output_image = image.copy()  # Make a copy to draw on

# --- Define class names (modify as needed) ---
class_names = [
    "person", "bicycle", "car", "motorbike", "aeroplane",
    "bus", "train", "truck", "boat", "traffic light"
]

# --- Generate a distinct color for each class (BGR format) ---
colors = [tuple(np.random.randint(0, 256, 3).tolist()) for _ in range(10)]

# --- Process each detection ---
for i in range(len(bboxes)):
    # Get final detection values.
    box = bboxes[i]             # [x1, y1, x2, y2]
    score = scores[i]           # detection confidence
    cls_id = int(class_ids[i])  # class index
    mask = masks[i]             # (160, 160) mask

    # Choose a color based on the class id.
    color = colors[cls_id % len(colors)]
    
    # --- Resize the mask to match the source image dimensions ---
    # (In case the source image size is different from 160x160.)
    mask = mask.astype(np.float32)
    mask_resized = cv2.resize(mask, (output_image.shape[1], output_image.shape[0]))

    # --- Threshold the mask to obtain a binary mask ---
    mask_bin = (mask_resized > 0.5).astype(np.uint8)

    # --- Create a colored mask overlay ---
    colored_mask = np.zeros_like(output_image, dtype=np.uint8)
    colored_mask[mask_bin == 1] = color

    # Blend the colored mask with the image
    alpha = 0.5  # Transparency factor
    output_image = cv2.addWeighted(output_image, 1.0, colored_mask, alpha, 0)

    # --- Draw the bounding box ---
    x1, y1, x2, y2 = map(int, box)
    cv2.rectangle(output_image, (x1, y1), (x2, y2), color, 2)

    # --- Prepare and draw the label text (class name and score) ---
    label = f"{class_names[cls_id]}: {score:.2f}"
    (text_w, text_h), baseline = cv2.getTextSize(label, cv2.FONT_HERSHEY_SIMPLEX, 0.5, 1)
    cv2.rectangle(output_image, (x1, y1 - text_h - baseline), (x1 + text_w, y1), color, thickness=-1)
    cv2.putText(output_image, label, (x1, y1 - baseline), cv2.FONT_HERSHEY_SIMPLEX,
                0.5, (255, 255, 255), thickness=1)

# --- Display the final image ---
cv2.imshow("Detections and Segmentation", output_image)
cv2.waitKey(0)
cv2.destroyAllWindows()

# --- Save the output image ---
cv2.imwrite("output_new.jpg", output_image)
```

I've tried the above code but I think the bbox, class_id, scores, masks are not extracted as expected.

**Comment by Y-T-G** [comment]
I am not sure why you're starting from 1

`torch.max(class_prob[:, :, 1:], dim=2)`
